math/src/fft/fft_inputs.rs

```diff
@@ -0,0 +1,76 @@
+use super::StarkField;
+use crate::FieldElement;
+
+// FFTINPUTS TRAIT
+// ================================================================================================
+
+/// Defines the shape of the input passed as input for FFT computation.
+pub trait FftInputs<B: StarkField> {
+    /// Returns the length of this fftinputs.
+    fn len(&self) -> usize;
+
+    /// Combines the result of smaller discrete fourier transforms into a larger DFT.
+    fn butterfly(&mut self, offset: usize, stride: usize);
+
+    /// Combines the result of smaller discrete fourier transforms multiplied with a
+    /// twiddle factor into a larger DFT.
+    fn butterfly_twiddle(&mut self, twiddle: B, offset: usize, stride: usize);
+
+    /// Swaps the element at index i with the element at index j.
+    fn swap(&mut self, i: usize, j: usize);
+
+    /// Helper method used to replace each element in fftinputs with respective `poly`
+    /// item times `init_offset`, where the `init_offset` gets updated by `offset_factor`
+    /// right after each element update.
+    ///
+    /// NOTE: When there's no `poly`, the item in fftinputs are updated with their product
+    /// with `init_offset` and `init_offset` gets updated by `offset_factor` after the update.
+    fn shift_by(&mut self, poly: Option<&Self>, offset_factor: B, init_offset: &mut B);
```

Comment:
> I would probably not have `poly` parameter here and instead have another function which takes this extra parameter as non-optional.

I will introduce this new function in the next PR where we would need to implement `chunk_mut` in `FftInputs` to handle `poly_evaluation_with_offset`.

>Also, for init_offset we should probably just use type B as whether it is mutable or not is an implementation detail.

>Lastly, maybe flip the order of parameters so that init_offset comes before offset_factor.

I've actioned on them in the commit cdacf86.